### PR TITLE
WSTEAMA-369 Migrate MostRead From Legacy part of Codebase

### PR DESCRIPTION
**Pull request by louisearchibald**
Resolves 
- https://jira.dev.bbc.co.uk/browse/WSTEAMA-369
- https://jira.dev.bbc.co.uk/browse/WSTEAMA-371
- https://jira.dev.bbc.co.uk/browse/WSTEAMA-375

Overall changes
======
Migrating the MostReadContainer from it's current location in `src/app/legacy/containers/MostRead` to a new location at `src/app/components/MostRead`

Code changes
======

- TypeScript-ifying all files
- Updating imports and references
- Simplify the Most Read props, to just include the props which are required for rendering the component
- Include the component on the new HomePage by hooking it into the Curation component 
- Ensuring existing tests pass

Testing
======

Helpful Links
======

[Coding Standards](https://github.com/bbc/simorgh/blob/latest/docs/Coding-Standards/README.md)

[Repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

### CODE REVIEW

File: src/app/components/Curation/getComponentName/index.ts [hunk 1]
```diff
@@ -10,14 +10,16 @@ export const COMPONENT_NAMES = {
   SIMPLE_CURATION_GRID: 'simple-curation-grid',
   HIERARCHICAL_CURATION_GRID: 'hierarchical-curation-grid',
   NOT_SUPPORTED: 'not-supported',
+  MOST_READ: 'most-read',
 } as const;
 
-const { NONE, BANNER, COLLECTION } = VISUAL_STYLE;
+const { NONE, BANNER, COLLECTION, RANKED } = VISUAL_STYLE;
```
Comment (karinathomasbbc) [inline src/app/components/Curation/getComponentName/index.ts:16]: We will need tests for this scenario

File: src/app/components/Curation/index.tsx [hunk 2]
```diff
@@ -1,4 +1,5 @@
 /** @jsx jsx */
+/* @jsxFrag React.Fragment */
```
Comment (karinathomasbbc) [inline src/app/components/Curation/index.tsx]: I think this line can be removed, as we are no longer rendering the most read inside a `<>`

File: src/app/components/Curation/index.tsx [hunk 3]
```diff
@@ -43,9 +46,8 @@ const Curation = ({
   headingLevel = 2,
   position = 0,
   curationLength = 0,
+  mostRead,
 }: CurationProps) => {
-  if (!promos.length) return null;
```
Comment (karinathomasbbc) [inline src/app/components/Curation/index.tsx:47]: I think we should move this logic to the curation grid section below - we don't want to display a header if there is a curation with an undefined or empty summaries array. We should be able to see this locally - there are multiple curation headers, with nothing rendered below.

File: src/app/components/MostRead/Canonical/Item/__snapshots__/index.test.tsx.snap
Comment (karinathomasbbc) [inline src/app/components/MostRead/Canonical/Item/__snapshots__/index.test.tsx.snap:1]: **Note to Reviewers:** (and in general about all of the snapshots in this PR): 

We made the decision to copy and paste the Most Read component from the legacy container. All snapshots & tests were copied, and we have done due diligence to ensure that all the tests are still passing where applicable (some tests have been removed, as they were no longer relevant). 

Although these snapshots appear as new, we have meticulously checked that they match the legacy snapshots as close as possible - with the exception of class renaming, or reduction of multi-line into a single line. 

We also have some visual (Chromatic) snapshots to cater for the various scenarios, so this will ensure that the Most Read component is as it was before


Screenshot 2023-05-17 at 12 29 43
Screenshot 2023-05-17 at 12 37 16

File: src/app/components/MostRead/README.md
Comment (karinathomasbbc) [inline src/app/components/MostRead/README.md:1]: Is there an existing README for the legacy component?

File: src/app/components/MostRead/index.test.tsx [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+        pageType: FRONT_PAGE,
+      },
+      {
+        description: 'should not render most read amp on home page',
+        service: 'mundo',
+        mostReadToggle: true,
+        isAmp: true,
+        variant: null,
+        renderExpectation: shouldRenderMostReadAmp,
```
Comment (karinathomasbbc) [inline src/app/components/MostRead/index.test.tsx:216]: Let's demo what the Most Read component looks like on AMP on the HomePage - we might need to either change the wording of this test, or the expectation based on the outcome

File: src/app/components/MostRead/index.tsx [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+          backgroundColor={backgroundColour}
+        />
+        <Amp
+          endpoint={`${process.env.SIMORGH_MOST_READ_CDN_URL}${endpoint}`}
+          size={size}
+        />
+      </MostReadSection>
+    ) : null;
+
+  // Do not render on Canonical if data is not provided
```
Comment (karinathomasbbc) [inline src/app/components/MostRead/index.tsx:69]: **Note to Reviewers:** This is a change from the existing behaviour of the legacy container, where the container would fetch the data on the client side. We have decided that the new canonical component will not do any client side data fetching, and will just render the data which has been passed to it. 

The BFF already includes the mostRead response for CPS Assets, Articles and HomePages, so it should be fairly straightforward to get this working once we migrate all pages types to the new component

File: src/app/components/MostRead/utilities/generateCSPHash/index.test.ts
Comment (karinathomasbbc) [inline src/app/components/MostRead/utilities/generateCSPHash/index.test.ts:1]: **Note to Reviewers:**  this file, and a number of the other files within the utilities directory have been copied from the legacy component and converted to typescript. So, for the moment, there is duplication between the two components, however this will be cleaned up in a future PR

File: src/app/components/ThemeProvider/fontSizes.ts [hunk 6]
```diff
@@ -1,13 +1,13 @@
...
+import { GROUP_B_ONLY, GROUP_D_MIN_WIDTH } from './fontMediaQueries';
 import { GelFontSize, TypographyScript } from '../../models/types/theming';
 
 const getGelFontSizeStyles =
   (size: GelFontSize) => (script: TypographyScript) => ({
     fontSize: `${pixelsToRem(script[size].groupA.fontSize)}rem`,
     lineHeight: `${pixelsToRem(script[size].groupA.lineHeight)}rem`,
 
-    [GROUP_B_MIN_WIDTH]: {
+    [GROUP_B_ONLY]: {
```
Comment (karinathomasbbc) [inline src/app/components/ThemeProvider/fontSizes.ts:10]: **Note to Reviewers:** We found this bug during snapshot testing - the new snapshots did not match the old snapshots due to the incorrect MQ being used here. This will have affected a number of other snapshots - mainly for any new TS components  

This is to replicate the functionality in https://github.com/bbc/simorgh/blob/f80390e6b26e6e60a2ab082ac94917e41b87e410/src/app/legacy/psammead/gel-foundations/src/getTypeSizes.js#L25-L26

File: src/app/components/ThemeProvider/spacings.ts [hunk 7]
```diff
@@ -1,9 +1,16 @@
 import pixelsToRem from '../../utilities/pixelsToRem';
 
+/** 0.5rem (8px) */
```
Comment (karinathomasbbc) [inline src/app/components/ThemeProvider/spacings.ts:3]: **Note to Reviewers:** This (and other comments) was to assist with our migration, and documenting the instructions on how to standardise components: https://paper.dropbox.com/doc/Migrate-MostRead-from-Legacy-0FHywUTsf55Brd9VaYUxs

File: src/app/legacy/containers/MostRead/Canonical/Item/__snapshots__/index.test.jsx.snap [hunk 8]
```diff
@@ -161,8 +161,8 @@ exports[`MostReadItemWrapper should render ltr correctly with 10 items 1`] = `
 }
 
 .emotion-7 {
-  font-family: ReithSerif,Helvetica,Arial,sans-serif;
-  font-weight: 300;
+  font-family: Helmet,Freesans,Helvetica,Arial,sans-serif;
```
Comment (karinathomasbbc) [inline src/app/legacy/containers/MostRead/Canonical/Item/__snapshots__/index.test.jsx.snap:164]: **Note to Reviewers:** We had to tidy up some of the legacy tests to ensure that the new tests would also work - the legacy snapshots were updated, and then copied to the new component for consistency

File: src/app/lib/utilities/filterPopularStaleData/index.js [hunk 9]
```diff
@@ -4,26 +4,27 @@ import {
...
-  // The ARES test endpoint for most read/watched renders fixture data, so the data is stale
-  const isTest = process.env.SIMORGH_APP_ENV === 'test';
+
   const logEvent =
     popularType === 'mostRead' ? MOST_READ_STALE_DATA : MOST_WATCHED_STALE_DATA;
   // Do not show most read if lastRecordUpdated is greater than 60min as this means PopAPI has failed twice
   // in succession. This suggests ATI may be having issues, hence risk of stale data.
-  if (!isTest && isDataStale(data.lastRecordTimeStamp)) {
+
+  if (isLive() && isDataStale(data.lastRecordTimeStamp)) {
```
Comment (karinathomasbbc) [inline src/app/lib/utilities/filterPopularStaleData/index.js]: We can possibly revert these changes now, as we are no longer checking for stale data in the new component

File: src/app/routes/homePage/getInitialData/index.ts [hunk 10]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
```
Comment (karinathomasbbc) [inline src/app/routes/homePage/getInitialData/index.ts]: Can this and the lint rule below be removed?

File: src/app/components/MostRead/README.md
Comment (louisearchibald) [reply src/app/components/MostRead/README.md:1]: Not that I could find! So I'd popped in this template so we wouldn't forget about writing one up. I'll have another comb through legacy to see if there's one hiding, but there's certainly not one inside the existing MostRead folder.

File: src/app/lib/utilities/filterPopularStaleData/index.js (hunk 9 above)
Comment (karinathomasbbc) [reply src/app/lib/utilities/filterPopularStaleData/index.js]: This change has caused tests to fail - now reverted

File: src/app/components/MostRead/README.md
Comment (karinathomasbbc) [reply src/app/components/MostRead/README.md:1]: Let's tackle this as part of the Component Health ticket https://jira.dev.bbc.co.uk/browse/WSTEAMA-376

**Review by karinathomasbbc — APPROVED** [review]

File: src/app/components/Curation/getComponentName/index.ts (hunk 1 above)
Comment (karinathomasbbc) [reply src/app/components/Curation/getComponentName/index.ts:16]: This wil be part of ticket https://jira.dev.bbc.co.uk/browse/WSTEAMA-371

File: src/app/lib/config/services/archive.ts [hunk 11]
```diff
@@ -3,6 +3,7 @@ import '#psammead/psammead-locales/moment/en-gb';
 import '#psammead/moment-timezone-include/tz/Europe/London';
 import withContext from '../../../contexts/utils/withContext';
 import { DefaultServiceConfig } from '../../../models/types/serviceConfig';
+// import { number } from '@storybook/addon-knobs';
```
Comment (karinathomasbbc) [inline src/app/lib/config/services/archive.ts]: Let's delete this

File: src/app/legacy/containers/SocialEmbed/Cps/__snapshots__/index.test.jsx.snap
Comment (karinathomasbbc) [inline src/app/legacy/containers/SocialEmbed/Cps/__snapshots__/index.test.jsx.snap:1]: **Note to reviewers:** See https://github.com/bbc/simorgh/pull/10817#discussion_r1196670161

File: ws-nextjs-app/pages/[service]/new_live/[id]/__snapshots__/live.test.tsx.snap
Comment (karinathomasbbc) [inline ws-nextjs-app/pages/[service]/new_live/[id]/__snapshots__/live.test.tsx.snap:1]: **Note to reviewers:** See https://github.com/bbc/simorgh/pull/10817#discussion_r1196670161

File: src/app/legacy/containers/SocialEmbed/__snapshots__/index.test.jsx.snap
Comment (karinathomasbbc) [inline src/app/legacy/containers/SocialEmbed/__snapshots__/index.test.jsx.snap:1]: **Note to reviewers:** See https://github.com/bbc/simorgh/pull/10817#discussion_r1196670161

File: src/app/legacy/psammead/psammead-social-embed/src/__snapshots__/index.test.jsx.snap
Comment (karinathomasbbc) [inline src/app/legacy/psammead/psammead-social-embed/src/__snapshots__/index.test.jsx.snap:1]: **Note to reviewers:** See https://github.com/bbc/simorgh/pull/10817#discussion_r1196670161

File: src/app/components/MostRead/index.test.tsx (hunk 4 above)
Comment (karinathomasbbc) [reply src/app/components/MostRead/index.test.tsx:216]: I've raised https://jira.dev.bbc.co.uk/browse/WSTEAMA-405 to capture this

File: src/app/routes/homePage/getInitialData/index.ts (hunk 10 above)
Comment (louisearchibald) [reply src/app/routes/homePage/getInitialData/index.ts]: I have removed the `no-explicit-any` lint rule, but removing the other causes a problem further down the file as it's using an `@ts-expect-error` so have left that one in.

Comment (karinathomasbbc) [reply src/app/routes/homePage/getInitialData/index.ts]: The ban-comment lint rule is only required if there's a `ts-ignore`. Can we replace `ts-ignore` with `ts-expect-error`?

File: src/app/components/MostRead/index.tsx (hunk 5 above)
Comment (HarveyPeachey) [reply src/app/components/MostRead/index.tsx:69]: What was the reason for client side fetching? Was it to do with content freshness?

Comment (karinathomasbbc) [reply src/app/components/MostRead/index.tsx:69]: I believe it was partly to do with content freshness, and also (I am guessing) something to do with how we used to fetch the data - before the BFF, we only fetched the content from ARES via the .json endpoint, but then the MostRead component would be responsible for fetching its own data? Although I could be completely wrong - @andrewscfc or @amoore108 would you happen to know more?

File: src/app/components/MostRead/README.md [hunk 12]
```diff
@@ -0,0 +1,17 @@
+# MostRead
```
Comment (HarveyPeachey) [inline src/app/components/MostRead/README.md:1]: Are we planning to fill out the README at this stage?

File: src/app/components/MostRead/Label/index.test.tsx [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+import MostReadSectionLabel from '.';
+
+describe('MostReadSectionLabel assertion', () => {
+  it('should render most-read section label with correct attributes', async () => {
+    const { container } = render(
+      <ServiceContextProvider service="persian">
+        <MostReadSectionLabel />
+      </ServiceContextProvider>,
+    );
+    const label = container.getElementsByTagName('span')[2];
```
Comment (HarveyPeachey) [inline src/app/components/MostRead/Label/index.test.tsx:13]: Not sure the scope of this migration, but would be nice to see these built in JS methods swapped out for testing library queries https://testing-library.com/docs/queries/about/#types-of-queries. But if it's not in the scope no need to worry about it

File: src/app/components/MostRead/Canonical/Item/index.tsx [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+export const MostReadItemWrapper = React.forwardRef(
+  (
+    { dir, children, columnLayout }: PropsWithChildren<MostReadItemProps>,
+    ref,
+  ) => (
+    // @ts-expect-error: Legacy grid expects `children` to be passed as props. However, due to coding best practices, we must nest children between the opening and closing tags
+    <Grid
+      css={styles.grid}
+      {...mostReadItemGridProps(columnLayout)}
+      parentColumns={getParentColumns(columnLayout)} // parentColumns is required here because on IE, this component would be rendered before it's parent therefore not receiving the parent's grid columns values so we have to explicitly pass it as a prop here so it works on IE
```
Comment (HarveyPeachey) [inline src/app/components/MostRead/Canonical/Item/index.tsx:75]: Do we need to worry about IE support anymore?

**Review by HarveyPeachey — APPROVED** [review]
Amazing work! I've worked on the most read component way back when, and I know how much of pain it can be, so well done for doing this everybody involved!

File: src/app/components/MostRead/index.tsx (hunk 5 above)
Comment (HarveyPeachey) [reply src/app/components/MostRead/index.tsx:69]: Okay, yeah I guess it depends if it was reliant on content freshness, now that we're relying on the BFF for this data has the content freshness been affected? And was that a major factor for the front pages?

File: src/app/components/MostRead/Canonical/Item/index.tsx (hunk 14 above)
Comment (karinathomasbbc) [reply src/app/components/MostRead/Canonical/Item/index.tsx:75]: Hmm good question, probably not, but we have just lifted and shifted this component - psammead grid and all (🤦 ), so I think we should probably leave this as is for now?

File: src/app/components/MostRead/README.md (hunk 12 above)
Comment (karinathomasbbc) [reply src/app/components/MostRead/README.md:1]: We have another ticket relating to component health, so we were planning to tackle it as part of that.

File: src/app/components/MostRead/Canonical/Item/index.tsx (hunk 14 above)
Comment (pvaliani) [reply src/app/components/MostRead/Canonical/Item/index.tsx:75]: According to our latest Browser Support I.E is Level C accounting for less than 0.1% and we’ll only investigate and make a call if content is completely unavailable

File: src/app/routes/homePage/getInitialData/index.ts [hunk 15]
```diff
@@ -37,7 +36,7 @@ export default async ({
 
     const optHeaders = { 'ctx-service-env': env };
 
-    // @ts-ignore - Ignore fetchPageData argument types
+    // @ts-expect-error - Ignore fetchPageData argument types
```
Comment (pvaliani) [inline src/app/routes/homePage/getInitialData/index.ts:39]: Are we keeping a note/log of the `@ts-expect-error` annotations and `@ts-ignore` to patch these up at a later date in a follow up ticket?

File: src/app/components/MostRead/index.tsx (hunk 5 above)
Comment (pvaliani) [reply src/app/components/MostRead/index.tsx:69]: Regarding content freshness via the BFF, there is currently a ticket to investigate the delay in time to publish content based on updates to TIPO which could affect `MostRead` but probably more pertinent to `Live` pages:

https://jira.dev.bbc.co.uk/browse/WSTEAMA-378?focusedCommentId=5945427&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-5945427

File: src/app/routes/homePage/getInitialData/index.ts (hunk 15 above)
Comment (karinathomasbbc) [reply src/app/routes/homePage/getInitialData/index.ts:39]: We can't always fix them... we have chosen to use `ts-expect-error` over `ts-ignore` (see https://paper.dropbox.com/doc/Migrate-MostRead-from-Legacy--B43g~QNaxpavx9j5UKF1MoreAg-0FHywUTsf55Brd9VaYUxs#:uid=470449461357824298478247&h2=Use-@ts-expect-error-instead-o), because you need to supply a reason. 

If something that is annotated with `ts-expect-error` gets fixed by something else, and it's not an error anymore, then type-checking will kick in and force us to remove the annotation.

Ideally we should not have a lot of `ts-expect-errors` in our production code, it is more often used in tests for error handling, or testing with partial data / incomplete props etc. 

But that said, there would be no harm in having some type of ticket in our tech backlog to come and assess / revisit / audit these every once in a while?

File: src/app/components/MostRead/Label/index.test.tsx (hunk 13 above)
Comment (pvaliani) [reply src/app/components/MostRead/Label/index.test.tsx:13]: Yep, agree. A lot of our newer tests have been written this way and I recall agreeing with JR to avoid using JS methods unless necessary. Whilst it doesn't mention it explicitly in our standards (https://github.com/bbc/simorgh/blob/latest/docs/Coding-Standards/Testing.stories.mdx), the examples in there use library queries instead of JS methods.

**Review by pvaliani — APPROVED** [review]
Awesome work. Had a good play around locally. I suggest that it would be good to also do a swarm code review on this piece of work as I know a lot of great knowledge has been built as a result 🥳

File: src/app/routes/homePage/getInitialData/index.ts (hunk 15 above)
Comment (pvaliani) [reply src/app/routes/homePage/getInitialData/index.ts:39]: Awesome, good work with using `ts-expect-error` sounds ideal!
